### ISSUE TITLE
Add Tweedie distributions to scipy.stats

### ISSUE THREAD

**Issue by lorentzenchr** [scipy/scipy#11291]
**Is your feature request related to a problem? Please describe.**
The [Compound Poisson Gamma distribution](https://en.wikipedia.org/wiki/Compound_Poisson_distribution#Compound_Poisson_Gamma_distribution) is often encountered in statistical modelling. It corresponds to a [Tweedie distribution](https://en.wikipedia.org/wiki/Tweedie_distribution) with power parameter between 1 and 2.
It would be nice to estimate the power parameter with MLE and to (efficiently) sample from these distributions.

**Describe the solution you'd like**
A new distribution in the stats submodule, e.g. `scipy.stats.tweedie(x, power=, ..)`

**Describe alternatives you've considered**
Sampling can be done by sampling from Poisson and Gamma distributions. There are also alternatives for MLE of the power parameter, see https://arxiv.org/abs/1609.03297, but those are also not easily implemented. An alternative is to use the [tweedie package](https://cran.r-project.org/web/packages/tweedie/index.html) from R. 

**References**
The hard part of an implementation is Wright's generalized Bessel function, see https://dlmf.nist.gov/10.46.E1.

**Comment by rlucas7** [comment]
@lorentzenchr thanks for your interest. 

I think we're usually open to adding new distributions which are commonly used and Tweedie would match those criteria. If someone were to implement the Tweedie they'd need to implement a `cdf`, `pdf` methods in the class as well as `rvs` and usually a moment or `stats` method although I don't think that the moments/stats is a hard requirement. 
On what is needed it may be helpful to look at the developer docs for adding a distribution:
http://scipy.github.io/devdocs/dev/contributor/adding_new.html#adding-new
and the general dev docs (dev setup etc):
http://scipy.github.io/devdocs/dev/contributor/contributor_toc.html

Second thing is that the arxiv paper you reference discusses a Tweedie ***regression*** (which uses a Tweedie ***distribution***) and not the distribution itself. If you are interested in implementing a Tweedie ***regression*** that would usually be done over in the `statsmodels` package rather than here in SciPy.

Are you interested in implementing the Tweedie distribution?

**Comment by josef-pkt** [comment]
statsmodels has Tweedie as a family in GLM for regression.

The problem with Tweedie is that the likelihood function pdf has no closed form expression. There are approximations, but the full, exact pdf has an infinite sum AFAIR.
I never looked for or at the cdf of Tweedie.

**Comment by lorentzenchr** [comment]
@rlucas7 Thanks for your reply. This issue is about the Tweedie distribution, not Tweedie regression, although the main usage may turn out to be support of Tweedie regression in downstream packages like statsmodels, scikit-learn, pymc3, etc.

I could try to give it a try in pure python, based on [Series evaluation of Tweedie exponential dispersion models densities](https://eprints.usq.edu.au/2335/1/Dunn_Smyth_Stats_and_Comp_v15n4.pdf) (https://doi.org/10.1007/s11222-005-4070-y).
I see 2 problems:
- For 1<p<2, there is a point mass at zero. How to deal with this for an otherwise continuous random variable?
- Wright's generalized Bessel function needs to be evaluated for the pdf. Should this function be exposed publicly in `scipy.special`? Maybe as separate PR?

@josef-pkt The exponential is an infinite sum, too:smirk:

**Comment by rlucas7** [comment]
>@rlucas7 Thanks for your reply. This issue is about the Tweedie distribution, not Tweedie regression, although the main usage may turn out to be support of Tweedie regression in downstream packages like statsmodels, scikit-learn, pymc3, etc.

Ok, then in scope SciPy. 
If the main usage is for downstream models it sounds like `statsmodels` already has something. @josef-pkt are there benefits or synergies here if distribution is implemented in Scipy?

I'm not sure the sole application is in regression models, I've used in in soils, horticultural, and oenology work and in my previous job in insurance (somewhat in former and extensively in the latter). There are uses in risk management outside of the regression case and in soils and horticulture outside regression.

>I could try to give it a try in pure python, based on Series evaluation of Tweedie exponential dispersion models densities (https://doi.org/10.1007/s11222-005-4070-y).

IIUC that paper ***does not*** implement a Wright generalized Bessel function, rather they approximate the function via one of several ways, e.g. series truncation, Fourier (numerical) inversion via FFT, I think one of Dunn or Smyth's papers also does a saddlepoint approximation (which isn't good for density approximations if `p` is close to 1). While all those approaches may be useful to have, on a first implementation I'd say pick 1 approach and implement.

>For 1<p<2, there is a point mass at zero. How to deal with this for an otherwise continuous random variable?

For univariate distributions there are 3 classes to inherit from, `rv_generic`, `rv_continuous`, and `rv_discrete`. The Tweedie type of distribution would best fit in the `rv_generic`, in general any distribution with a mixture of point mass and absolute continuity is a mixed distribution and fits into the `rv_generic`. Unfortunately, there aren't any of those distributions currently implemented (not that I'm aware if anyway so it might make sense to post this question on the scipy-dev email list too to get feedback from the other developers about how to approach. In particular, some of the requirements for implementation (e.g. `rvs`, `cdf`, `pdf` etc.) may be less stringent (or more) and would be good to get consensus there before you invest more efforts in approach.

>Wright's generalized Bessel function needs to be evaluated for the pdf. Should this function be exposed publicly in scipy.special? Maybe as separate PR?

There are several cases of `cdf` or `pdf` implementations in `scipy.special` so that might be a good first step, implement the cdf or pdf as the wright generalized bessel function in special, then once that is merged implement `tweedie` in `stats` and call the function from special. You would probably need to implement an `rvs` method still and that would need either a `cdf` inversion (numerically) or some other specialized methods (once it gets to that point you could determine the best approach). 

Either way though I'd recommend to post on the scipy-dev email list and get feedback from the rest of the stats folks to make sure they are in agreement. Some of them might have better ideas that I'm not thinking of at the moment which might help you so please start a thread there describing what you'd like to do and if people agree with approach and if they have recommendations. 

Hope that helps.

**Comment by josef-pkt** [comment]
> are there benefits or synergies here if distribution is implemented in Scipy

Good implementation of distributions is always useful, especially if "special" experts help.

about statsmodels implementation:
For GLM regression, we actually don't need the likelihood function to estimate the mean parameters using iteratively reweighted least squares (IRLS). The power and variance parameters are estimated using some other methods.
We have an approximate loglikelihood function (based on saddlepoint approximation IIRC), but I think it's not very accurate especially for the variance and power parameters. 
One of the maintainers/contributors also had another implementation which is also just approximate, IIRC.

(I also have some older code for poisson mixtures,  compound Poisson, including fft versions, but again it is difficult to find approximation errors and the truncation choices for the general case. )

**Comment by tmcclintock** [comment]
Any update on this? I'm interested in having access to this distribution as well. Also thank you all for the informative discussion above!

**Comment by lorentzenchr** [comment]
@tmcclintock #11313 was merged and gives access to the loglikelihood, e.g. with statsmodels class Tweedie, see [#8281](https://github.com/statsmodels/statsmodels/pull/8281). Another independent alternative is [glum.TweedieDistribution.log_likelihood](https://glum.readthedocs.io/en/latest/glm.html#glum.TweedieDistribution.log_likelihood).

In the [scipy mailing list thread](https://mail.python.org/pipermail/scipy-dev/2020-March/024074.html), it became clear that Tweedie does not fit into the current API of scipy because the point mass at zero would require a mix of `rv_continuous` and `rv_discrete`.

**Comment by lucascolley** [comment]
Thanks for the update @lorentzenchr, I'll close this now. For anyone reading this, these distributions won't be included in SciPy any time soon, but you can find implementations in other libraries as linked above 👍

**Comment by tmcclintock** [comment]
@lorentzenchr great, thank you!
